### PR TITLE
Fix prepare + apply

### PR DESCRIPTION
**Pull request by jmamou**

### CODE REVIEW

**Comment by jmamou** [comment]
- simplify suppress_tokens
- refactor AssistantToTargetTranslator to avoid moving tensors to cpu
- fix _prepare_assistant_input_ids of USD
- fix logits_processors bug: logits_processors was called after sampling assistant token ids.

**Comment by gauravjain14** [comment]
@jmamou 
To try this, do I need to apply the changes on top of https://github.com/keyboardAnt/transformers/pull/6?

**Comment by jmamou** [comment]
> @jmamou To try this, do I need to apply the changes on top of #6?

no, just checkout `fix_prepare` branch

File: src/transformers/generation/candidate_generator.py [hunk 1]
```diff
@@ -577,22 +577,21 @@ def __init__(self, target_tokenizer: "PreTrainedTokenizerBase", assistant_tokeni
...
-        return {
-            assistant_vocab[tok]: target_vocab[tok] for tok in set(target_vocab.keys()) & set(assistant_vocab.keys())
-        }
-
+        max_assistant_index = max(assistant_vocab.values())
+        assistant_to_target_input_ids = torch.full((max_assistant_index+1,), -1, dtype=int) # -1 means not in target vocab
+        for tok, idx in assistant_vocab.items():
+            if tok in target_vocab:
+                assistant_to_target_input_ids[idx] = target_vocab[tok]
+        return assistant_to_target_input_ids
```
Comment (gauravjain14) [inline src/transformers/generation/candidate_generator.py:588]: Do we expect this to have an impact on the end-to-end accuracy or this is lossless?

**Review by gauravjain14 — APPROVED** [review]
Overall, the changes look good to me. I was able to run the failing test cases and they seem to have been resolved with this.

File: src/transformers/generation/candidate_generator.py (hunk 1 above)
Comment (jmamou) [reply src/transformers/generation/candidate_generator.py:588]: @gauravjain14 it is lossless

File: src/transformers/generation/candidate_generator.py [hunk 2]
```diff
@@ -732,15 +735,17 @@ def get_candidates(self, input_ids: torch.LongTensor) -> Tuple[torch.LongTensor,
...
         assistant_output = self.assistant_model.generate(**generation_args, **self.assistant_kwargs)
         self.assistant_kwargs["past_key_values"] = assistant_output.past_key_values
 
         candidate_logits = torch.stack(assistant_output.scores, dim=1)
 
         # Use translator to convert tokens and logits
         candidate_ids = assistant_output.sequences
-        candidate_logits = self._atm_translator.logits_processors(input_ids=candidate_ids, scores=candidate_logits)
+        self._prev_assistant_ids = candidate_ids
         target_ids = self._atm_translator.get_target_ids(assistant_input_ids, target_input_ids, candidate_ids)
```
Comment (keyboardAnt) [inline src/transformers/generation/candidate_generator.py:747]: Wdyt about removing `candidate_ids`?
```suggestion
        self._prev_assistant_ids = assistant_output.sequences
        target_ids = self._atm_translator.get_target_ids(assistant_input_ids, target_input_ids, self._prev_assistant_ids)
```

File: src/transformers/generation/candidate_generator.py [hunk 3]
```diff
@@ -732,15 +735,17 @@ def get_candidates(self, input_ids: torch.LongTensor) -> Tuple[torch.LongTensor,
...
+        self._candidates_target_seq_len = target_ids.shape[-1]
```
Comment (keyboardAnt) [inline src/transformers/generation/candidate_generator.py:748]: Here, `self._candidates_target_seq_len` is set to the length of all the given sequence rather than the length of only the candidates, right? If so, the name `self._candidates_target_seq_len` might not be most informative, and even confusing. Please consider using the original name (`self._prev_target_seq_len`) or propose a better alternative.

File: src/transformers/generation/candidate_generator.py [hunk 4]
```diff
@@ -708,8 +710,9 @@ def __init__(
             logits_processor,
         )
         # Track sequence lengths and previous assistant IDs
-        self._prev_target_seq_len: int = 0
+        self._candidates_target_seq_len: int = 0
```
Comment (keyboardAnt) [inline src/transformers/generation/candidate_generator.py:713]: Please see my comment below about this new name.

File: src/transformers/generation/candidate_generator.py [hunk 5]
```diff
@@ -765,11 +772,15 @@ def _prepare_assistant_input_ids(self, target_input_ids: torch.LongTensor) -> to
...
+            assistant_input_ids = assistant_new_ids
         else:
-            self._prev_assistant_ids = torch.cat([self._prev_assistant_ids, assistant_new_ids], dim=-1)
-
-        return self._prev_assistant_ids
+            i = self._candidates_target_seq_len+1-target_seq_len
+            if i > 0:
+                self._prev_assistant_ids = self._prev_assistant_ids[:,:-i]
+            assistant_input_ids = torch.cat([self._prev_assistant_ids, assistant_new_ids], dim=-1)      
+        assistant_input_ids = assistant_input_ids.to(torch.int)
```
Comment (keyboardAnt) [inline src/transformers/generation/candidate_generator.py:781]: According to the documentation, `cat` operates on arrays of the same type. Wdyt about ensuring that `self._prev_assistant_ids` and `assistant_new_ids` are already of `torch.int` type?

File: src/transformers/generation/candidate_generator.py [hunk 6]
```diff
@@ -732,15 +735,17 @@ def get_candidates(self, input_ids: torch.LongTensor) -> Tuple[torch.LongTensor,
...
+        generation_args['logits_processor'] = self._atm_translator.logits_processors
         assistant_output = self.assistant_model.generate(**generation_args, **self.assistant_kwargs)
         self.assistant_kwargs["past_key_values"] = assistant_output.past_key_values
 
         candidate_logits = torch.stack(assistant_output.scores, dim=1)
 
         # Use translator to convert tokens and logits
         candidate_ids = assistant_output.sequences
-        candidate_logits = self._atm_translator.logits_processors(input_ids=candidate_ids, scores=candidate_logits)
+        self._prev_assistant_ids = candidate_ids
```
Comment (keyboardAnt) [inline src/transformers/generation/candidate_generator.py:746]: Wouldn't it be easier to set `self._prev_assistant_ids` to be the output of `self._prepare_assistant_input_ids(target_input_ids)` so that the function `_prepare_assistant_input_ids` doesn't have to crop out its suffix?

File: src/transformers/generation/candidate_generator.py [hunk 7]
```diff
@@ -765,11 +772,15 @@ def _prepare_assistant_input_ids(self, target_input_ids: torch.LongTensor) -> to
...
         if self._prev_assistant_ids is None:
-            self._prev_assistant_ids = assistant_new_ids
+            assistant_input_ids = assistant_new_ids
         else:
-            self._prev_assistant_ids = torch.cat([self._prev_assistant_ids, assistant_new_ids], dim=-1)
-
-        return self._prev_assistant_ids
+            i = self._candidates_target_seq_len+1-target_seq_len
+            if i > 0:
+                self._prev_assistant_ids = self._prev_assistant_ids[:,:-i]
```
Comment (keyboardAnt) [inline src/transformers/generation/candidate_generator.py:779]: I believe we should strive to simplify this part in later PRs.

**Review by keyboardAnt — COMMENTED** [review]
Thanks @jmamou! It's good news that @gauravjain14's tests pass for this PR. I added some questions and minor comments, mostly about simplifying the implementation.

File: src/transformers/generation/utils.py [hunk 8]
```diff
@@ -858,6 +858,7 @@ def _get_candidate_generator(
                         logits_processor=logits_processor,
                         target_tokenizer=target_tokenizer,
                         assistant_tokenizer=assistant_tokenizer,
+                        target_vocab_size=self.config.vocab_size # required in the case that self.config.vocab_size is different from the length of target_tokenizer.get_vocab()
```
[the remark this answers is not in the document]
Comment (jmamou) [reply src/transformers/generation/utils.py:861]: `microsoft/Phi-3-medium-128k-instruct`

**Comment by jmamou** [comment]
I just though about another change, I will check it tomorrow and ping when ready

File: src/transformers/generation/candidate_generator.py (hunk 3 above)
Comment (jmamou) [reply src/transformers/generation/candidate_generator.py:748]: `self._candidates_target_seq_len` is that length of target seq length including draft candidates before validation. 'self._prev_target_seq_len' length of target seq length including validated tokens from previous spec iteration. Renamed `self._candidates_target_seq_len` to `self._target_seq_len_with_candidates`

File: src/transformers/generation/candidate_generator.py (hunk 5 above)
Comment (jmamou) [reply src/transformers/generation/candidate_generator.py:781]: do you mean adding before `cat`
```
self._prev_assistant_ids = self._prev_assistant_ids.to(torch.int)
assistant_new_ids = assistant_new_ids.to(torch.int)
```

File: src/transformers/generation/candidate_generator.py (hunk 6 above)
Comment (jmamou) [reply src/transformers/generation/candidate_generator.py:746]: `self._prev_assistant_ids` includes also the candidates not yet validated while in `self._prepare_assistant_input_ids` we crop unvalidated draft tokens.

File: src/transformers/generation/candidate_generator.py (hunk 5 above)
Comment (keyboardAnt) [reply src/transformers/generation/candidate_generator.py:781]: Wdyt about ensuring we only assign `torch.int` to `self._prev_assistant_ids` and `assistant_new_ids` in the first place—so that we never need to cast them into `torch.int`?

File: src/transformers/generation/candidate_generator.py (hunk 6 above)
Comment (keyboardAnt) [reply src/transformers/generation/candidate_generator.py]: Okay. Perhaps we can simplify this in later PRs.

File: src/transformers/generation/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (keyboardAnt) [reply src/transformers/generation/utils.py:861]: Why do we need `self.config.vocab_size` in the first place? 👀

**Review by keyboardAnt — COMMENTED** [review]
I'm somewhat puzzled by the `target_vocab_size` argument :eyes:

File: src/transformers/generation/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jmamou) [reply src/transformers/generation/utils.py:861]: we need it for padding in `AssistantToTargetTranslator`

File: src/transformers/generation/candidate_generator.py (hunk 5 above)
Comment (jmamou) [reply src/transformers/generation/candidate_generator.py:781]: we get all the IDs from the tokenizer and their type is `int`. Do you think that it is necessary to ensure they are of `int` type?

File: src/transformers/generation/candidate_generator.py [hunk 9]
```diff
@@ -565,34 +564,38 @@ class AssistantToTargetTranslator:
...
+    def __init__(self, target_tokenizer: "PreTrainedTokenizerBase", assistant_tokenizer: "PreTrainedTokenizerBase", assistant_model_device, target_vocab_size: int, 
+                 assistant_vocab_size:int):
         self._target_tokenizer: "PreTrainedTokenizerBase" = target_tokenizer
         self._assistant_tokenizer: "PreTrainedTokenizerBase" = assistant_tokenizer
-        self._assistant_to_target_input_ids: dict[int, int] = self._get_assistant_to_target_input_ids()
-        self.suppress_input_ids: list[int] = self._get_suppress_input_ids()
+        self.assistant_model_device = assistant_model_device
+        target_tokenizer_vocab_size: int = len(target_tokenizer.get_vocab())
+        # paddind is required in the case that target_vocab_size is bigger than set(target_tokenizer.get_vocab().keys())
+        if target_tokenizer_vocab_size < target_vocab_size:
```
[the remark this answers is not in the document]
Comment (jmamou) [reply src/transformers/generation/candidate_generator.py:574]: with model `microsoft/Phi-3-medium-128k-instruct`
`len(target_tokenizer.get_vocab())` = 32011
while `config.vocab_size`= 32064

**Comment by jmamou** [comment]
> > with model `microsoft/Phi-3-medium-128k-instruct`
> > `len(target_tokenizer.get_vocab())` = 32011
> > while `config.vocab_size`= 32064
> 
> Where/why do we set `config.vocab_size = 32064` if we know that `len(target_tokenizer.get_vocab()) = 32011`?

we don't set it.
It is part of model config
https://huggingface.co/microsoft/Phi-3-medium-128k-instruct/blob/main/config.json#L169

File: src/transformers/generation/candidate_generator.py [hunk 10]
```diff
@@ -626,11 +626,11 @@ def get_target_logits(self, assistant_logits: torch.FloatTensor) -> torch.FloatT
 
         target_shape: tuple[int, ...] = (*assistant_logits.shape[:-1], self.target_vocab_size)
         target_logits: torch.FloatTensor = torch.full(target_shape, -float("inf")).to(self._assistant_model_device)
-        assistant_indices = self._assistant_to_target_input_ids != -1  # Mask for valid indices
-        target_indices = self._assistant_to_target_input_ids[assistant_indices]  # Exclude invalid indices
+        assistant_indices_mask = self._assistant_to_target_input_ids != -1  # Mask for valid indices
+        target_logits_supported_indices = self._assistant_to_target_input_ids[assistant_indices_mask]  # Exclude invalid indices
```
Comment (keyboardAnt) [inline src/transformers/generation/candidate_generator.py:630]: Please use standard comments (instead of inline comments)

**Review by keyboardAnt — CHANGES_REQUESTED** [review]
@jmamou, thanks for the clarification about the vocabulary sizes. I’ve added a few comments. My main concern is the suggested breaking change in `SuppressTokensLogitsProcessor`.

File: src/transformers/generation/logits_process.py [hunk 11]
```diff
@@ -1860,15 +1860,19 @@ class SuppressTokensLogitsProcessor(LogitsProcessor):
     ```
     """
 
-    def __init__(self, suppress_tokens, device: str = "cpu"):
-        self.suppress_tokens = torch.tensor(list(suppress_tokens), device=device)
+    def __init__(
+        self, mapped_tokens, assistant_vocab_size, assistant_model_device, filter_value: float = -float("Inf")
```
[the remark this answers is not in the document]
Comment (jmamou) [reply src/transformers/generation/logits_process.py:1864]: It is not just renaming.
we don't need explicitly suppress_tokens. mapped_tokens is more informative.

**Comment by jmamou** [comment]
> @jmamou, thanks for the clarification about the vocabulary sizes. I’ve added a few comments. My main concern is the suggested breaking change in `SuppressTokensLogitsProcessor`.

original implementation was buggy and not optimal. please explain your concern ...

**Comment by keyboardAnt** [comment]
> > @jmamou, thanks for the clarification about the vocabulary sizes. I’ve added a few comments. My main concern is the suggested breaking change in `SuppressTokensLogitsProcessor`.
> 
> 
> 
> original implementation of `SuppressTokensLogitsProcessor` was buggy and not optimal. please explain your concern ... 

My concern is that such a change might cause failures for users of Hugging Face Transformers who call `SuppressTokensLogitsProcessor` while expecting the existing API. Changing the API would require all users to adjust their current implementations.

Another option is to extend the API of the existing class without breaking it or to create an entirely new class.

**Comment by jmamou** [comment]
> SuppressTokensLogitsProcessor

Sorry for the misunderstanding ... I was not aware that `SuppressTokensLogitsProcessor` was already part of HF transformers.

**Comment by jmamou** [comment]
> > SuppressTokensLogitsProcessor
> 
> Sorry for the misunderstanding ... I was not aware that `SuppressTokensLogitsProcessor` was already part of HF transformers.

I opt for the second option of creating a new class.

**Comment by keyboardAnt** [comment]
> > > SuppressTokensLogitsProcessor
> 
> > 
> 
> > Sorry for the misunderstanding ... I was not aware that `SuppressTokensLogitsProcessor` was already part of HF transformers.
> 
> 
> 
> I opt for the second option of creating a new class.

Sounds good. Bugs in the existing `SuppressTokensLogitsProcessor` will then no longer be relevant for USD and can be reported to Hugging Face or fixed in separate PRs (not urgent).

**Comment by gauravjain14** [comment]
What is the expectation on the `generation_mode` being `ASSISTED_GENERATION` when speculative decoding with different tokenizers is enabled?

https://github.com/keyboardAnt/transformers/blob/1ed1de2fecdb5fe82f67e4fe742417f737370324/src/transformers/generation/utils.py#L2165

When I run this script - https://gist.github.com/gauravjain14/19edce088b1f1e7b5dc9ace684e53f8d - with `do_sample=True`
the first call into the function `generate` has `generation_mode=GenerationMode.ASSISTED_GENERATION` but subsequent calls into the function have `generation_mode=GenerationMode.GREEDY_SEARCH`.

Is this expected? @jmamou, @keyboardAnt?

**Comment by jmamou** [comment]
> What is the expectation on the `generation_mode` being `ASSISTED_GENERATION` when speculative decoding with different tokenizers is enabled?
> 
> https://github.com/keyboardAnt/transformers/blob/1ed1de2fecdb5fe82f67e4fe742417f737370324/src/transformers/generation/utils.py#L2165
> 
> When I run this script - https://gist.github.com/gauravjain14/19edce088b1f1e7b5dc9ace684e53f8d - with `do_sample=True` the first call into the function `generate` has `generation_mode=GenerationMode.ASSISTED_GENERATION` but subsequent calls into the function have `generation_mode=GenerationMode.GREEDY_SEARCH`.
> 
> Is this expected? @jmamou, @keyboardAnt?

`generation_mode` of the target is `GenerationMode.ASSISTED_GENERATION` while `generation_mode` of the assistant model should be `GenerationMode.SAMPLE` (`do_sample=True`) or `GenerationMode.GREEDY_SEARCH` (`do_sample=False`). 
That's the reason why you can get `GenerationMode.ASSISTED_GENERATION` for the first call to `generate` (`self` is target). But you should get `GenerationMode.SAMPLE` for subsequent `generate` calls (`self` is assistant) until the target `generate` call for validation.

**Comment by keyboardAnt** [comment]
@jmamou, please hit the 'Re-request Review' button when you're ready.

**Review by keyboardAnt — APPROVED** [review]
LGTM.

**Comment by keyboardAnt** [comment]
@jmamou, it [seems](https://github.com/huggingface/transformers/pull/35029/checks?check_run_id=34564392323) like the changes fail the CI tests. Do they pass for you locally?

**Comment by jmamou** [comment]
> @jmamou, it [seems](https://github.com/huggingface/transformers/pull/35029/checks?check_run_id=34564392323) like the changes fail the CI tests. Do they pass for you locally?

The failing tests don’t appear to be related to USD. Please note that these tests are for the current main branch version.
I hope that after rebasing USD onto the main branch, the tests will pass.

**Comment by jmamou** [comment]
> > @jmamou, it [seems](https://github.com/huggingface/transformers/pull/35029/checks?check_run_id=34564392323) like the changes fail the CI tests. Do they pass for you locally?
> 
> The failing tests don’t appear to be related to USD. Please note that these tests are for the current main branch version. I hope that after rebasing USD onto the main branch, the tests will pass.

I just resolved conflicts

**Comment by jmamou** [comment]
> @jmamou, it [seems](https://github.com/huggingface/transformers/pull/35029/checks?check_run_id=34564392323) like the changes fail the CI tests. Do they pass for you locally?

After solving conflicts and tests, remaining failing test does not seem to be related to USD https://app.circleci.com/pipelines/github/huggingface/transformers/113897/workflows/98283892-64b7-4e14-b8a3-8f7da8f9aa61/jobs/1523429?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary
